cudapoa/src/cudapoa_kernels.cuh

```diff
@@ -371,7 +342,347 @@ __launch_bounds__(GW_POA_KERNELS_MAX_THREADS_PER_BLOCK)
+                                                                 node_coverage_counts,
+                                                                 base_weights,
+                                                                 (sequence_begin_nodes_ids + s),
+                                                                 outgoing_edges_coverage,
+                                                                 outgoing_edges_coverage_count,
+                                                                 s,
+                                                                 max_sequences_per_poa,
+                                                                 max_nodes_per_graph);
+
+            if (error_code != 0)
+            {
+                consensus[0] = CUDAPOA_KERNEL_ERROR_ENCOUNTERED;
+                consensus[1] = error_code;
+                warp_error   = true;
+            }
+            else
+            {
+                sequence_lengths[0] = new_node_count;
+                // Run a topsort on the graph.
+#ifdef SPOA_ACCURATE
+                // Exactly matches racon CPU results
+                raconTopologicalSortDeviceUtil(sorted_poa,
+                                               node_id_to_pos,
+                                               new_node_count,
+                                               incoming_edge_count,
+                                               incoming_edges,
+                                               node_alignment_count,
+                                               node_alignments,
+                                               node_marks,
+                                               check_aligned_nodes,
+                                               nodes_to_visit,
+                                               (uint16_t)max_nodes_per_graph);
+#else
+                // Faster top sort
+                topologicalSortDeviceUtil(sorted_poa,
+                                          node_id_to_pos,
+                                          new_node_count,
+                                          incoming_edge_count,
+                                          outgoing_edges,
+                                          outgoing_edge_count,
+                                          sorted_poa_local_edge_count);
+#endif
+            }
+        }
+
+        __syncwarp();
+
+        warp_error = __shfl_sync(FULL_MASK, warp_error, 0);
+        if (warp_error)
+        {
+            return;
+        }
+    }
+}
+
+// Similar to generatePOAKernel(), but with a different launch bounds setting, used to call kernels using traceback buffer
+// i.e. for BM == BandMode::static_band_traceback or BM == BandMode::adaptive_band_traceback
+template <typename ScoreT, typename SizeT, typename TraceT, bool MSA = false, BandMode BM = BandMode::static_band_traceback>
+__launch_bounds__(MAX_THREADS_PER_BLOCK_72)
+    __global__ void generatePOAKernelTB(uint8_t* consensus_d,
```

Comment:
I think you're already explained this to me but I am not able to recall the exact reason now - if we were to combine the two kernels with templatization, is that also causing the register count to go up?

the main reason I'm asking is because a lot of the initial indexing code is copied over, and I'm concerned that they might go out of sync if one was updated

Reply:
I agree, it will be difficult to maintain the two copies of `generatePOAKernel()` and `generatePOAKernelTB()`. I had tried once and it added to the register count, but the code has changed since then, will retry.

Reply:
retry was successful! done.